coverage/historical/bitbucket/historical_coverage_report.py

```diff
@@ -0,0 +1,431 @@
+    difflines = diffstr.splitlines()
+    deleted_lines = []
+    i = 0
+    while i < len(difflines):
+        diff = difflines[i]
+        changed_blocks = diff_changes_pattern.findall(diff)
+        if not changed_blocks:
+            i += 1
+            continue
+        num_lines = int(changed_blocks[0])
+        for j in range(i+1, i+num_lines):
+            if difflines[j].startswith("-"):
+                diff_line = difflines[j][1:].strip()
+                if len(diff_line) < NUM_CHARS_DIFF:
+                    continue
+                deleted_lines.append(diff_line)
+        i += num_lines
+    return deleted_lines
+
+def line_exists_in_range(blame_line, line_number):
+    return blame_line['line_number'] == line_number
+
+def reconstruct_blame(commits, diffs, current_line, commit_cache):
+    """Reconstruct the blame using the diffs."""
+    for i, diff in enumerate(diffs):
+        if diff:
+            for line in diff.splitlines():
+                if line.startswith("+"):
+                    actual_line = line[1:].strip()
+                    if len(actual_line) < NUM_CHARS_DIFF:
+                        continue
+                    if actual_line == current_line:
+                        commit_author = get_commit_author(commits[i]['links']['self']['href'], commit_cache)
+                        return commit_author
+    return None
+def get_commits_for_file(workspace, repo_slug, file_path, commit_id):
+    """Fetch the list of commits for the file using the filehistory endpoint."""
+    headers = {
+        "Authorization": f"Basic {base64_encoded_credentials}"
+    }
+    commits = []
+    url = f"{BASE_URL}/repositories/{workspace}/{repo_slug}/filehistory/{commit_id}/{file_path}"
+    while url:
+        response = requests.get(url, headers=headers)
+        if not response.ok:
+            print(f"[get_commits_for_file] error in response = {response.text}")
+            return None
+        data = response.json()
+        commits.extend([value["commit"] for value in data["values"]])
+        url = data.get("next", None)
+    return commits
+
+
+def get_diff_for_commit(workspace, repo_slug, commit_id, filepath, diff_cache):
+    """Fetch the diff for a specific commit."""
+    headers = {
+        "Authorization": f"Basic {base64_encoded_credentials}"
+    }
+    cache_key = f'{commit_id}-{filepath}'
+    if cache_key in diff_cache:
+        return diff_cache[cache_key]
+    url = f"{BASE_URL}/repositories/{workspace}/{repo_slug}/diff/{commit_id}"
+    params = {"path": filepath}
+    response = requests.get(url, headers=headers, params=params)
+    if not response.ok:
+        print(f"No diff for commit: {commit_id}, file: {filepath}")
+        return None
+    diff = response.text
+    diff_cache[cache_key] = diff
+    return diff
+
+def get_commit_author(commit_url, commit_cache):
+    """
+    Get commit author
+    """
+    headers = {
+        "Authorization": f"Basic {base64_encoded_credentials}"
+    }
```

Comment:
You have added this Authorization header each time you initiate a request. This is not required and you just do this once in that whole file.